lib/srv/db/postgres/engine.go

```diff
@@ -52,6 +52,29 @@ type Engine struct {
 	Clock clockwork.Clock
 	// Log is used for logging.
 	Log logrus.FieldLogger
+	// client is a client connection.
+	client *pgproto3.Backend
+}
+
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.client = pgproto3.NewBackend(pgproto3.NewChunkReader(clientConn), clientConn)
+
+	// The proxy is supposed to pass a startup message it received from
+	// the psql client over to us, so wait for it and extract database
+	// and username from it.
+	err := e.handleStartup(e.client, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// SendError sends an error to connected client in a Postgres understandable format.
+func (e *Engine) SendError(err error) {
+	if err := e.client.Send(toErrorResponse(err)); err != nil && !utils.IsOKNetworkError(err) {
```

Comment:
I'd probably leave error handling to the caller now that this is a separate interface method instead of suppressing it here.

Reply:
@r0mant I can do that, but currently each implementation has a bit different condition on when an error is being send. That's why I left it here. Do you remember what is the reason? Or that just on oversight?

Reply:
@jakule Yeah, I don't think there was any particular reason other than those were done at different times so turned out a bit inconsistent unfortunately. It's not big deal either way.